README.md
WPS-FM2TaskBar
==============
FM/2 Taskbar is a small utility program that works in conjunction with the WPS Window List, program objects and Launchpad to control what's running on your system and where it runs.
 
 LICENSE
===============
- GNU GPL V2

COMPILE TOOLS
===============
* 

AUTHORS
===============
* Mark Kimes
* Steven H. Levine

LINKS
===============
* 

docs/History.Txt

TASKBAR Change Log
==================

Fri Oct 31 23:10:24 PDT 2003 Steven H. Levine <steve53@earthlink    >

	* Release Version - 1.8

Tue Jun 03 19:13:25 PDT 2003 Steven H. Levine <steve53@earthlink    >

	* Release Version - 1.7d

	* Correct switchlist icon background redraw - 2nd try

Fri May 30 20:09:59 PDT 2003 Steven H. Levine <steve53@earthlink    >

	* Release Version - 1.7c

	* Correct switchlist icon background redraw

Wed May 28 14:55:39 PDT 2003 Steven H. Levine <steve53@earthlink    >

	* Release Version - 1.7b

	* Restore WPS to taskbar selection list

Sat Apr 19 23:51:56 PDT 2003 Steven H. Levine <steve53@earthlink    >

	* Release Version - 1.7a

	* Correct popup

        * Correct icon background fill

Tue Sep 25 12:20:36 PDT 2001 Steven H. Levine <steve53@earthlink    >

	* Release Version - 1.7

	* Build with VAC 3.08

$Id: History.Txt,v 1.3 2003/11/01 07:36:30 root Exp $


MENU.C
#define INCL_DOS
#define INCL_WIN

#include <os2.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "taskbar.h"

/* strip all leading spaces and tabs from a string */
#define lstrip(s)         strip_lead_char(" \t",(s))

/* strip all trailing spaces and tabs from a string */
#define rstrip(s)         strip_trail_char(" \t",(s))

/*strip all trailing linefeeds and carriage returns from a string */
#define stripcr(s)        strip_trail_char("\r\n",(s))

#pragma alloc_text(MENU,strip_trail_char,strip_lead_char,FreeStartData)
#pragma alloc_text(MENU,FreeMenu,SubmenuLoad,MenuLoad,MenuSave)


CHAR * strip_trail_char (register CHAR *strip,register CHAR *a) {

  register CHAR *p;

  if(a && *a && strip && *strip) {
    p = a + (strlen(a) - 1);
    while (*a && strchr(strip,*p) != NULL) {
      *p = 0;
      p--;
    }
  }
  return a;
}


CHAR * strip_lead_char (register CHAR *strip,register CHAR *a) {

  register CHAR *p = a;

  if(a && *a && strip && *strip) {
    while(*p && strchr(strip,*p) != NULL)
      p++;
    if(p != a)
      memmove(a,p,strlen(p) + 1);
  }
  return a;
}


VOID FreeStartData (STARTDATA *sd) {

  if(sd) {
    if(sd->PgmTitle)
      free(sd->PgmTitle);
    if(sd->PgmName)
      free(sd->PgmName);
    if(sd->PgmInputs)
      free(sd->PgmInputs);
    if(sd->Environment)
      free(sd->Environment);
    if(sd->IconFile)
      free(sd->IconFile);
    if(sd->ObjectBuffer)
      free(sd->ObjectBuffer);
    free(sd);
  }
}


VOID FreeMenu (MENUSTUFF *menu) {

  MENUSTUFF *next,*info;

  if(!menu)
    menu = menuhead;
  info = menu;
  while(info) {
    next = info->next;
    if(info->type == MIS_SUBMENU && info->sub)
      FreeMenu(info->sub);
    if(info->sd)
      FreeStartData(info->sd);
    free(info);
    info = next;
  }
}


VOID SubmenuLoad (MENUSTUFF *head,FILE *fp) {

  MENUSTUFF  *info;
  CHAR        s[1024],*p;
  INT         state;

  while(!feof(fp)) {
    if(!fgets(s,1024,fp))
      break;
    s[1023] = 0;
    stripcr(s);
    lstrip(rstrip(s));
    if(!*s || *s == ';')
      continue;
    if(!strcmp(s,"SUBMENU END"))
      break;
    if(!strcmp(s,"MENUITEM"))
      state = 0;
    else if(!strcmp(s,"MENUCONTROL"))
      state = MIS_BREAK;
    else
      state = MIS_SUBMENU;
    /* get item title */
    if(!fgets(s,1023,fp))
      break;
    stripcr(s);
    lstrip(rstrip(s));
    if(!strncmp(s,"Title:",6)) {
      p += 6;
      while(*p == ' ')
        p++;
      if(*p) {

      }
    }
  }
}


MENUSTUFF * MenuLoad (CHAR *filename) {

  MENUSTUFF  *head = NULL;
  FILE       *fp;

  fp = fopen(filename,"r");
  if(fp) {
    head = malloc(sizeof(MENUSTUFF));
    if(head) {
      memset(head,0,sizeof(MENUSTUFF));
      SubmenuLoad(head,fp);
    }
    fclose(fp);
  }
  return head;
}


BOOL MenuSave (MENUSTUFF *menu,CHAR *filename) {

  MENUSTUFF *info;
  FILE      *fp;
  BOOL       ret = FALSE;

  fp = fopen(filename,"w");
  if(fp) {
    if(!menu)
      menu = menuhead;
    info = menu;
    while(info) {

      info = info->next;
    }
    fclose(fp);
  }
  return ret;
}



FINDWPS.C
#define INCL_DOS
#define INCL_WIN

#include <os2.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "taskbar.h"


VOID FindDesktop (HAB hab) {

  CHAR  Class[24];
  HENUM henum;
  TID   tid;

  /* Get to bottommost window handle of the desktop */
  hwndBottom = WinQueryWindow(HWND_DESKTOP, QW_BOTTOM);
  /* Enumerate all windows at desktop z-order */
  henum = WinBeginEnumWindows(hwndBottom);
  hwndWPS = (HWND)0;
  while((hwndWPS = WinGetNextWindow(henum)) != NULLHANDLE) {
    /* Now get the class name of that window handle */
    WinQueryClassName(hwndWPS,sizeof(Class),(PCH)Class);
    if(!strcmp(Class,"#37"))          /* desktop class */
      break;
  }
  WinEndEnumWindows(henum);

  /* Without the WPS installed we can only get the desktop window handle */
  hwndDesktop = WinQueryDesktopWindow(hab,NULLHANDLE);

  /* find the PID of the desktop process */
  WinQueryWindowProcess((hwndWPS) ? hwndWPS : (hwndDesktop) ?
                        hwndDesktop : hwndBottom,&dtPid,&tid);
}


PROCSTAT.H
#pragma linkage (DosQProcStatus, far16 pascal)
USHORT DosQProcStatus (PVOID pvBuf, USHORT cbBuf);

#define PROCESS_END_INDICATOR   3       // Indicates end of process structs

#pragma pack(1)


typedef struct _SUMMARY {
  ULONG   ulThreadCount;              // Number of threads in system
  ULONG   ulProcessCount;             // Number of processes in system
  ULONG   ulModuleCount;              // Number of modules in system
} SUMMARY, *PSUMMARY;

typedef struct _THREADINFO {
  ULONG   ulRecType;                  // Record type (thread = 100)
  USHORT  tidWithinProcess;           // TID within process (TID is 4 bytes!!)
  USHORT  usSlot;                     // Unique thread slot number
  ULONG   ulBlockId;                  // Sleep id thread is sleeping on
  ULONG   ulPriority;                 // Priority
  ULONG   ulSysTime;                  // Thread System Time
  ULONG   ulUserTime;                 // Thread User Time
  UCHAR   uchState;                   // 1=ready,2=blocked,5=running
  UCHAR   uchPad;                     // Filler
  USHORT  usPad;                      // Filler
} THREADINFO, *PTHREADINFO;

typedef struct _PROCESSINFO {
  ULONG       ulEndIndicator;         // 1 means not end, 3 means last entry
  PTHREADINFO ptiFirst;               // Address of the 1st Thread Control Blk
  USHORT      pid;                    // Process ID (2 bytes - PID is 4 bytes)
  USHORT      pidParent;              // Parent's process ID
  ULONG       ulType;                 // Process Type
  ULONG       ulStatus;               // Process Status
  ULONG       idSession;              // Session ID
  USHORT      hModRef;                // Module handle of EXE
  USHORT      usThreadCount;          // Number of threads in this process
  ULONG       ulReserved;             // Unknown
  PVOID       pvReserved;             // Unknown
  USHORT      usSem16Count;           // Number of 16-bit system semaphores
  USHORT      usDllCount;             // Number of Dlls used by process
  USHORT      usShrMemHandles;        // Number of shared memory handles
  USHORT      usReserved;             // Unknown
  PUSHORT     pusSem16TableAddr;      // Address of a 16-bit semaphore table
  PUSHORT     pusDllTableAddr;        // Address of a Dll table
  PUSHORT     pusShrMemTableAddr;     // Address of a shared memory table
} PROCESSINFO, *PPROCESSINFO;

typedef struct _SEMINFO {
  struct _SEMINFO *pNext;             // Ptr to next block (NULL on last one)
  UINT     idOwningThread;            // ID of owning thread?
  UCHAR    fbFlags;                   // Semaphore flags
  UCHAR    uchReferenceCount;         // Number of references
  UCHAR    uchRequestCount;           // Number of requests
  UCHAR    uchReserved;               // Unknown
  ULONG    ulReserved;                // Unknown
  UINT     uiReserved;                // Unknown
  CHAR     szSemName[ 1 ];            // ASCIIZ semaphore name
} SEMINFO, *PSEMINFO;

typedef struct _SHRMEMINFO {
  struct _SHRMEMINFO *pNext;          // Ptr to next block (NULL on last one)
  USHORT      usMemHandle;            // Shared memory handle (?)
  SEL         selMem;                 // Selector
  USHORT      usReferenceCount;       // Number of references
  CHAR        szMemName[ 1 ];         // ASCIIZ shared memory name
} SHRMEMINFO, *PSHRMEMINFO;

typedef struct _MODINFO {
  struct _MODINFO *pNext;             // Ptr to next block (NULL on last one)
  USHORT   hMod;                      // Module handle
  USHORT   usModType;                 // Module type (0=16bit,1=32bit)
  ULONG    ulModRefCount;             // Count of module references
  ULONG    ulSegmentCount;            // Number of segments in module
  ULONG    ulDontKnow1;               //
  PSZ      szModName;                 // Addr of fully qualified module name
  USHORT   usModRef[ 1 ];             // Handles of module references
} MODINFO, *PMODINFO;

typedef struct _BUFFHEADER {
  PSUMMARY        psumm;              // SUMMARY section ptr
  PPROCESSINFO    ppi;                // PROCESS section ptr
  PSEMINFO        psi;                // SEM section ptr (add 16 to offset)
  PVOID           pDontKnow1;         //
  PSHRMEMINFO     psmi;               // SHARED MEMORY section ptr
  PMODINFO        pmi;                // MODULE section ptr
  PVOID           pDontKnow2;         //
  PVOID           pDontKnow3;         //
} BUFFHEADER, *PBUFFHEADER;


#pragma pack()


TASKHOOK.C
/****************************************************************************
    taskhook.c  - input hook DLL

    FM/2 Taskbar is a WPS enhancement
    copyright (c) 2001 by Mark Kimes

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Revisions	19 Aug 01 MK - Release
		25 Sep 01 SHL - Add missing #includes

*****************************************************************************/


#define INCL_DOS
#define INCL_WIN

#include <os2.h>

#include <string.h>

#include "taskbar.h"

#pragma alloc_text(ONCE,InitDLL,StartInputHook,StartSendHook,StopHooks)

typedef struct HOOKED {
  HWND           hwnd;
  PFNWP          oldproc;
  struct HOOKED *next;
} HOOKED;

/***********************************************************************/
/*  Global variables.                                                  */
/***********************************************************************/
HAB     habDLL;
HWND    hwndTaskbar,hwndDesktop,hwndWPS,hwndBottom;
HMODULE hMod;
PFN     pfnInput,pfnSend;
BOOL    fWinShowing,fMenuShowing,fSlidingFocus,fNoZorderChange,
        fB2Bottom,fB3Close,fLeft = TRUE,fRight,fCenter = TRUE,
        fDlgShowing,fB3DblClk,fKeyShowing,fNoHot,fLeftHot,fWrapMouse,
        fCtrlTab = TRUE,fStayUp;
LONG    xScreen,yScreen,xIcon,yIcon,yNext,yHi = 4;
PID     dtPid,myPid;


#define KILLMSG()  {pqMsg->hwnd = NULLHANDLE; return(TRUE);}


BOOL PostMsg (HWND h, ULONG msg, MPARAM mp1, MPARAM mp2) {

  BOOL rc = WinPostMsg(h,msg,mp1,mp2);

  if(!rc) {

    PIB *ppib;
    TIB *ptib;

    if(!DosGetInfoBlocks(&ptib,&ppib)) {

      PID pid;
      TID tid;

      if(WinQueryWindowProcess(h,&pid,&tid)) {
        if(pid != ppib->pib_ulpid || tid != ptib->tib_ptib2->tib2_ultid) {
          for(;;) {
            DosSleep(1L);
            rc = WinPostMsg(h,msg,mp1,mp2);
            if(!rc) {
              if(!WinIsWindow((HAB)0,h))
                break;
            }
            else
              break;
          }
        }
      }
    }
  }
  return rc;
}


/***********************************************************************/
/*  InitDLL: This function sets up the DLL and sets all variables      */
/***********************************************************************/
BOOL EXPENTRY InitDLL (HAB hab) {

  habDLL = hab;

  yIcon = WinQuerySysValue(HWND_DESKTOP,SV_CYICON);

  if(!DosLoadModule(NULL,0,"TASKHOOK",&hMod) &&
     !DosQueryProcAddr(hMod,4,"InputProc",&pfnInput) &&
     !DosQueryProcAddr(hMod,34,"SendProc",&pfnSend))
    return TRUE;
  return FALSE;
}


/***********************************************************************/
/*  StartInputHook: This function starts the hook filtering.           */
/***********************************************************************/
BOOL EXPENTRY StartInputHook (void) {

  return WinSetHook(habDLL,NULLHANDLE,HK_INPUT,pfnInput,hMod);
}


/***********************************************************************/
/*  StartSendHook: This function starts the hook filtering.            */
/***********************************************************************/
BOOL EXPENTRY StartSendHook (void) {

  return WinSetHook(habDLL,NULLHANDLE,HK_SENDMSG,pfnSend,hMod);
}


/***********************************************************************/
/*  StopHooks: This function stops the hook filtering.             */
/***********************************************************************/
BOOL EXPENTRY StopHooks (void) {

  if(WinReleaseHook(habDLL,NULLHANDLE,HK_INPUT,pfnInput,hMod) &&
     (!pfnSend || WinReleaseHook(habDLL,NULLHANDLE,HK_SENDMSG,pfnSend,hMod)) &&
     !DosFreeModule(hMod))
    return TRUE;
  return FALSE;
}


/***********************************************************************/
/*  InputProc: This is the input filter routine.                       */
/*  While the hook is active, all messages come here                   */
/*  before being dispatched.                                           */
/***********************************************************************/
BOOL EXPENTRY InputProc (HAB hab, PQMSG pqMsg, ULONG fs) {

  /* mouse is captured -- ignore msg */
  if(WinQueryCapture(HWND_DESKTOP) != NULLHANDLE)
    return FALSE;
  switch(pqMsg->msg) {
    case WM_MOUSEMOVE:
      switch(fWinShowing) {
        case FALSE:
          if(!fDlgShowing && !fKeyShowing) {
            if(pqMsg->ptl.y < yHi && ((fLeft && fRight && fCenter) ||
               (fLeft && pqMsg->ptl.x < (xScreen / 4)) ||
               (fRight && pqMsg->ptl.x > (xScreen / 4) * 3) ||
               (fCenter && (pqMsg->ptl.x > (xScreen / 4) - 4) &&
                            pqMsg->ptl.x < ((xScreen / 4) * 3) + 4))) {
              fWinShowing = (BOOL)WinSendMsg(hwndTaskbar,UM_SHOW,
                                             MPVOID,MPVOID);
              break;
            }
            if(fWrapMouse && (pqMsg->ptl.x <= 0 ||
                              pqMsg->ptl.x >= xScreen - 1 ||
                              pqMsg->ptl.y <= 0 ||
                              pqMsg->ptl.y >= yScreen - 1)) {
              if(pqMsg->ptl.x <= 0)
                pqMsg->ptl.x = xScreen - 2;
              else if(pqMsg->ptl.x >= xScreen - 1)
                pqMsg->ptl.x = 1;
              else if(pqMsg->ptl.y <= 0)
                pqMsg->ptl.y = yScreen - 2;
              else if(pqMsg->ptl.y >= yScreen - 1)
                pqMsg->ptl.y = 1;
              WinSetPointerPos(HWND_DESKTOP,pqMsg->ptl.x,pqMsg->ptl.y);
              break;
            }
          }
          /*                                                                                      *\
           * If enabled, here we catch all mouse movements, to set the window under the mouse     *
           * pointer as the active one, if it isn't currently active or the window list or        *
           * optionally the Desktop window.                                                       *
          \*                                                                                      */
          /* If enabled, use sliding focus to activate window
             under the mouse pointer (with some exceptions).
             Caution! Menus have a class WC_MENU, but their
             parent is not the frame window WC_FRAME but the
             Desktop itself. */
          while(fSlidingFocus) {

            static UCHAR    ucClassname[7];     /* Window class f.e. #1 for WC_FRAME */
            static UCHAR    ucWindowText[33];   /* Window name f.e. OS/2 2.0 Desktop */
            static HWND     hwndActive;         /* Window handle of active frame class window on Desktop */
            static HWND     hwndApplication;    /* Window handle of application under mouse pointer */
            static HWND     hwndTitlebar;
                                                /* Window handle of applications parent window */
            static HWND     hwndApplicationParent;
            static HWND     hwndTop;            /* Window handle on top of zorder */
            BOOL            ctrldown = FALSE;
            HENUM           henum;

            if(WinGetKeyState(HWND_DESKTOP, VK_SHIFT) & 0x8000)
                return(FALSE);
            if(WinGetKeyState(HWND_DESKTOP, VK_CTRL) & 0x8000)
                ctrldown = TRUE;
            /* Query the currently active window, where HWND_DESKTOP
               is the parent window. It will be a WC_FRAME class
               window */
            hwndActive = WinQueryActiveWindow(HWND_DESKTOP);
            if(fNoZorderChange && ctrldown) {
              henum = WinBeginEnumWindows(HWND_DESKTOP);
              hwndTop = WinGetNextWindow(henum);
              WinEndEnumWindows(henum);
            }
            else
              hwndTop = (HWND)0;
            WinQueryWindowText(hwndActive,sizeof(ucWindowText),
                               ucWindowText);
            /* Don't switch away from the WC_FRAME class tasklist */
            if(!strcmp(ucWindowText,"Window List"))
              break;
            /* Get message target window */
            hwndApplication = pqMsg->hwnd;
            /* If the window under the mouse pointer is one of the
               Desktops, don't do any changes */
            if((hwndApplication == hwndDesktop) ||
               (hwndApplication == hwndWPS))
              break;
            /* Get parent window of current window */
            hwndApplicationParent = WinQueryWindow(hwndApplication,
                                                   QW_PARENT);
            /* Loop until we get the Desktop window handle. The
               previous child window of the Desktop is then the
               WC_FRAME class window of the point under the mouse
               pointer which is not the Desktop. */
            while(hwndApplicationParent != hwndDesktop) {
              hwndApplication = hwndApplicationParent;
              hwndApplicationParent = WinQueryWindow(hwndApplication,
                                                     QW_PARENT);
            }
            /* Query the class of the frame window of the
               designated target of WM_MOUSEMOVE */
            WinQueryClassName(hwndApplication, sizeof(ucClassname),
                              ucClassname);
            /* Don't switch to menu windows */
            if(!strcmp(ucClassname,"#4"))
              return(FALSE);
            /* Don't switch to a combobox's listbox windows */
            if(!strcmp(ucClassname,"#7"))
              return(FALSE);
            /* Query the frame window name of the designated
               target of WM_MOUSEMOVE */
            WinQueryWindowText(hwndApplication, sizeof(ucWindowText),
                               ucWindowText);
            /* Don't switch to seamless Win-OS2 menus */
            if(strstr(ucWindowText,"Seamless"))
              return(FALSE);
            /* Sort with expected descending probability, to avoid
               unnecessary cpu load */
            for (;;) {
              /* Don't switch if previous windows equals current one */
              if(hwndActive == hwndApplication) {
                if(hwndTop == hwndApplication || !fNoZorderChange ||
                   !ctrldown)
                break;
              }
              /* Only switch to WC_FRAME class windows */
              if(strcmp(ucClassname,"#1"))
                break;
              if(!ctrldown && fNoZorderChange) {
                /* Change focus, but preserve Z-order */
                /* Don't send WM_ACTIVATE to window with new focus */
                WinFocusChange(HWND_DESKTOP,WinWindowFromID(hwndApplication,
                               FID_CLIENT),FC_NOSETACTIVE);
                /* Activate new window */
                if((hwndTitlebar = WinWindowFromID(hwndApplication,
                                                   FID_TITLEBAR)) != (HWND)0)
                 PostMsg(hwndApplication, WM_ACTIVATE, MPFROMSHORT(TRUE),
                         MPFROMHWND(hwndTitlebar));
              }
              else if(ctrldown && fNoZorderChange)
                WinSetActiveWindow(HWND_DESKTOP,hwndApplication);
              else
                /* Switch to the new frame window. It will generate
                   all messages of deactivating old and activating
                   new frame window */
                WinSetFocus(HWND_DESKTOP,WinWindowFromID(hwndApplication,
                            FID_CLIENT));
              /* We changed the focus, don't pass this message to
                 the next hook in the chain */
              return(TRUE);
            }
            break;                              /* Exit loop now */
          }
          break;

        default:
          if(!fStayUp && !fDlgShowing && !fMenuShowing && !fKeyShowing &&
             pqMsg->ptl.y > yNext + yIcon + 4)
            fWinShowing = (BOOL)WinSendMsg(hwndTaskbar,UM_HIDE,MPVOID,MPVOID);
          break;
      }
      break;

    case WM_BUTTON2DOWN:
      if(!fDlgShowing && fB2Bottom) {

        static HWND hwndFrame;

        hwndFrame = WinQueryWindow(pqMsg->hwnd,QW_PARENT);
        /* If we click on a titlebar the current window handle
           equals the parent's titlebar window handle. If the
           Shift key is pressed we pass this message to the titlebar
           instead of processing to change its z-order */
        if((pqMsg->hwnd == WinWindowFromID(hwndFrame,FID_TITLEBAR)) &&
           (!(WinGetKeyState(HWND_DESKTOP,VK_SHIFT) & 0x8000))) {
          /* Set it to the bottom of all windows */
          WinSetWindowPos(hwndFrame,HWND_BOTTOM,0,0,0,0,
                          SWP_ZORDER | SWP_DEACTIVATE);
          KILLMSG();
        }
      }
      break;

    case WM_BUTTON3UP:
      if(fB3DblClk && !(WinGetKeyState(HWND_DESKTOP,VK_SHIFT) & 0x8000))
        KILLMSG();
      break;

    case WM_BUTTON3DOWN:
      if(!fDlgShowing && (fB3Close || fB3DblClk) &&
         !(WinGetKeyState(HWND_DESKTOP,VK_SHIFT) & 0x8000)) {

        static HWND hwndFrame;

        if(fB3Close) {
          hwndFrame = WinQueryWindow(pqMsg->hwnd,QW_PARENT);
          /* If we click on a titlebar the current window handle
             equals the parent's titlebar window handle. If the
             Shift key is pressed we pass this message to the titlebar
             instead of telling it to die. */
          if(pqMsg->hwnd == WinWindowFromID(hwndFrame,FID_TITLEBAR)) {
            /* close it */
            PostMsg(hwndFrame,WM_SYSCOMMAND,
                    MPFROM2SHORT(SC_CLOSE,0),MPVOID);
            KILLMSG();
            break;
          }
        }
        if(fB3DblClk) {

          static HWND hwndFrame;

          hwndFrame = WinQueryWindow(pqMsg->hwnd,QW_PARENT);
          /* If we click on a titlebar the current window handle
             equals the parent's titlebar window handle. If the
             Shift key is pressed we pass this message to the titlebar
             instead of processing to change its z-order */
          if(pqMsg->hwnd == WinWindowFromID(hwndFrame,FID_TITLEBAR)) {
//            PostMsg(pqMsg->hwnd,WM_BUTTON1DOWN,pqMsg->mp1,pqMsg->mp2);
            PostMsg(pqMsg->hwnd,WM_BUTTON1DBLCLK,pqMsg->mp1,pqMsg->mp2);
//            PostMsg(pqMsg->hwnd,WM_OPEN,pqMsg->mp1,pqMsg->mp2);
          }
          else {
            PostMsg(pqMsg->hwnd,WM_BUTTON1DOWN,pqMsg->mp1,pqMsg->mp2);
            PostMsg(pqMsg->hwnd,WM_BUTTON1UP,pqMsg->mp1,pqMsg->mp2);
            PostMsg(pqMsg->hwnd,WM_BUTTON1CLICK,pqMsg->mp1,pqMsg->mp2);
            PostMsg(pqMsg->hwnd,WM_SINGLESELECT,pqMsg->mp1,pqMsg->mp2);
            PostMsg(pqMsg->hwnd,WM_BUTTON1DBLCLK,pqMsg->mp1,pqMsg->mp2);
            PostMsg(pqMsg->hwnd,WM_OPEN,pqMsg->mp1,pqMsg->mp2);
            PostMsg(pqMsg->hwnd,WM_BUTTON1UP,pqMsg->mp1,pqMsg->mp2);
          }
          KILLMSG();
//          pqMsg->msg = WM_BUTTON1DBLCLK;  /* this sorta works */
        }
      }
      break;

    case WM_CHAR:
      if(!fNoHot && !fKeyShowing && !fDlgShowing) {

        USHORT usFlags    = SHORT1FROMMP(pqMsg->mp1);
        UCHAR  ucScanCode = CHAR4FROMMP(pqMsg->mp1);

        if(usFlags & KC_KEYUP)
          break;
        if((usFlags & (KC_SCANCODE | KC_CTRL | KC_ALT)) !=
           (KC_SCANCODE | KC_CTRL | KC_ALT))
          break;
        if((fCtrlTab && ucScanCode == 15) ||
          (fLeftHot && (ucScanCode == 44 || ucScanCode == 45)) ||
          (!fLeftHot && (ucScanCode == 51 || ucScanCode == 52))) {
          fKeyShowing = TRUE;
          if(!PostMsg(hwndTaskbar,UM_ALTTAB,MPFROMLONG((LONG)ucScanCode),
                      MPVOID))
            fKeyShowing = FALSE;
        }
      }
      break;

    case WM_TASKLIST:
      PostMsg(hwndTaskbar,UM_TASKLIST,pqMsg->mp1,pqMsg->mp2);
      break;
  }
  /* Pass the message on to the next hook in line. */
  return FALSE;
}


VOID EXPENTRY SendProc (HAB hab,PSMHSTRUCT psmh,BOOL fInterTask) {

  static UCHAR ucClassname[7];     /* Window class f.e. #1 for WC_FRAME */

  switch(psmh->msg) {
    case WM_CREATE:
      WinQueryClassName(psmh->hwnd,sizeof(ucClassname),ucClassname);
      if(!strcmp(ucClassname,"#1") || !strcmp(ucClassname,"Folder")) {

      }
      break;

    case WM_DESTROY:
      break;
  }
}


TASKBAR.H

/****************************************************************************

    FM/2 Taskbar Common Includes
    $Id: TASKBAR.H,v 1.7 2003/11/01 08:14:27 root Exp $

    Copyright (c) 1996, 2001 Mark Kimes
    Copyright (c) 2001, 2003 Steven Levine and Associates, Inc.

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Revisions	03 Apr 96 MK - Baseline
		19 Apr 03 SHL - Version 1.07a
		28 May 03 SHL - Version 1.07b
		30 May 03 SHL - Version 1.07c
		30 May 03 SHL - Version 1.07d
		31 Oct 03 SHL - Version 1.08

*****************************************************************************/

#define VERSION             "1.08"

#define MAIN_ID             1
#define MAIN_HELP           2
#define APP_POINTER         3
#define MAIN_TIMER          4
#define MAIN_DATETIME       5
#define MAIN_STARTUP        6
#define MAIN_VIRTUAL        7
#define BLANK_POINTER       8
#define MAIN_PROCS          9

#define START_TIMER         10
#define START_FRAME         11
#define START_BITMAP        12
#define START_ICON          13
#define FM2_ICON            14
#define START_VERSION       15

#define MAIN_FIRST          100

#define RUN_FRAME           200
#define RUN_FIND            201
#define RUN_NAME            202
#define RUN_DIR             203
#define RUN_ARGS            204
#define RUN_DIRHDR          205
#define RUN_ARGSHDR         206

#define SET_FRAME           300
#define SET_LEFT            301
#define SET_CENTER          302
#define SET_RIGHT           303
#define SET_HEIGHT          304
#define SET_NOZORDERCHANGE  305
#define SET_SLIDINGFOCUS    306
#define SET_REPOSMOUSE      307
#define SET_MOVEFOLDERS     308
#define SET_B2BOTTOM        309
#define SET_B3CLOSE         310
#define SET_B3DBLCLK        311
#define SET_UNDO            312
#define SET_DISABLEDESKTOPS 313
#define SET_ANIMATE         314
#define SET_NOHOT           315
#define SET_LEFTHOT         316
#define SET_WRAPMOUSE       317
#define SET_SHOWMINIWINS    318
#define SET_CTRLTAB         319
#define SET_PROCS           320

#define TAB_FRAME           400
#define TAB_ICON0           401
#define TAB_ICON1           402
#define TAB_ICON2           403
#define TAB_NAME            404

#define IDM_HELP            999
#define IDM_CLOSE           1000
#define IDM_SHOW            1001
#define IDM_HIDE            1002
#define IDM_KILL            1003
#define IDM_MOVEDESKTOP     1005
#define IDM_SETDATETIME     1007
#define IDM_WPSCLOCK        1008
#define IDM_ABOUT           1009
#define IDM_OPENFM2         1010
#define IDM_OPENFM2FOLDER   1011
#define IDM_OPENOS2WIN      1012
#define IDM_OPENOS2FULL     1013
#define IDM_OPENDOSWIN      1014
#define IDM_OPENDOSFULL     1015
#define IDM_OPENWINOS2      1016
#define IDM_OPENDRIVES      1017
#define IDM_OPENFONTPALETTE 1018
#define IDM_OPENSOLIDCOLOR  1019
#define IDM_OPENHIRESCOLOR  1020
#define IDM_OPENTEMPLATES   1021
#define IDM_OPENSETUP       1022
#define IDM_MINIMIZEALL     1026
#define IDM_RESTOREALL      1027
#define IDM_RESETDESKTOPS   1028
#define IDM_DEBUGINFO       1029
#define IDM_MOVEFOLDERS     1030
#define IDM_RUN             1031
#define IDM_SETUP           1032
#define IDM_OPENLAUNCHPAD   1034
#define IDM_OPENCLOCK       1036
#define IDM_RELOADTEXCLUDES 1037
#define IDM_RELOADVEXCLUDES 1038
#define IDM_EDITTEXCLUDES   1039
#define IDM_EDITVEXCLUDES   1040
#define IDM_LOGO            1045
#define IDM_REMOVEFROMLIST  1046
#define IDM_JUMPABLE        1047

#define MENU_FIRST          10000

#define UM_SHOW             (WM_USER + 1)
#define UM_HIDE             (WM_USER + 2)
#define UM_FOCUSME          (WM_USER + 3)
#define UM_MENUEND          (WM_USER + 4)
#define UM_TIMER            (WM_USER + 5)
#define UM_CONTEXTMENU      (WM_USER + 6)
#define UM_CLOSE            (WM_USER + 7)
#define UM_FREE             (WM_USER + 8)
#define UM_TEST             (WM_USER + 9)
#define UM_SETUP            (WM_USER + 10)
#define UM_MOVEDESKTOP      (WM_USER + 11)
#define UM_ALTTAB           (WM_USER + 12)
#define UM_GETLIST          (WM_USER + 13)
#define UM_PAINT            (WM_USER + 14)
#define UM_PROCS            (WM_USER + 15)
#define UM_TASKLIST         (WM_USER + 16)

#define WM_TASKLIST         0x00002f43

typedef struct SWENTRYLIST {
  SWENTRY             swe;
  USHORT              id;
  struct SWENTRYLIST *prev;
  struct SWENTRYLIST *next;
} SWENTRYLIST;

typedef struct {
  USHORT       size;
  USHORT       flags;
  PID          pid;
  ULONG        ulCount;
  SWENTRYLIST *swehead;
  SWENTRYLIST *swetail;
  HWND         hwndMenu;
  BOOL         killme;
} TASKDATA;

typedef struct {
  USHORT size;
  USHORT dummy;
  PID    pid;
  HWND   hwnd;
  ULONG  delay;
} KILLWIN;

typedef struct MENUSTUFF {
  USHORT            size;   // size of structure
  USHORT            type;   // MIS_* constants
  USHORT            id;     // id of menu item
  USHORT            dummy;
  STARTDATA        *sd;     // program to start, if any
  struct MENUSTUFF *next;   // points to next menu item, if any
  struct MENUSTUFF *prev;   // points to previous menu item, if any
  struct MENUSTUFF *parent; // points to parent menu, if any
  struct MENUSTUFF *sub;    // if type indicates submenu, this is it
} MENUSTUFF;

#ifdef DEFINE_GLOBALS
  #define DATADEF
#else
  #define DATADEF extern
#endif

DATADEF HPOINTER   hptrApp,hptrBlank;
DATADEF LONG       xVirtual,yVirtual;
DATADEF ULONG      ulAnimate;
DATADEF BOOL       fReposMouse,fMoveFolders,fExtravagent,fDisableDesktops,
                   fShowMiniwins,fProcs;
DATADEF HWND       hwndAppMenu;
DATADEF HINI       tbarprof;
DATADEF MENUSTUFF *menuhead;
DATADEF CHAR     **texclude,**vexclude;
DATADEF CHAR      *pgmname;

extern LONG        xScreen,yScreen,xIcon,yIcon,yNext,yHi;
extern PID         dtPid;
extern HWND        hwndDesktop,hwndWPS,hwndBottom,hwndTaskbar;
extern BOOL        fWinShowing,fMenuShowing,fSlidingFocus,fNoZorderChange,
                   fB2Bottom,fB3Close,fLeft,fRight,fCenter,fDlgShowing,
                   fB3DblClk,fKeyShowing,fNoHot,fLeftHot,fWrapMouse,
                   fCtrlTab,fStayUp;

/* findwps.c */
VOID FindDesktop (HAB hab);

/* The end */
